strawberryfields/tdm/tdmprogram.py

```diff
@@ -268,20 +387,139 @@ def __init__(self, N, name=None):
             self.N = N
         self.concurr_modes = sum(self.N)
 
-        self.total_timebins = 0
+        super().__init__(num_subsystems=self.concurr_modes, name=name)
+
+        self.type = "tdm"
+        self.timebins = 0
         self.spatial_modes = 0
         self.measured_modes = []
-        super().__init__(num_subsystems=self.concurr_modes, name=name)
+        self.rolled_circuit = None
+        # `unrolled_circuit` only contains the unrolled single-shot circuit
+        self.unrolled_circuit = None
+        self.run_options = {}
+        """dict[str, Any]: dictionary of default run options, to be passed to the engine upon
+        execution of the program. Note that if the ``run_options`` dictionary is passed
+        directly to :meth:`~.Engine.run`, it takes precedence over the run options specified
+        here.
+        """
 
     # pylint: disable=arguments-differ, invalid-overridden-method
-    def context(self, *args, copies=1, shift="default"):
-        input_check(args, copies)
-        self.copies = copies
+    def context(self, *args, shift="default"):
+        input_check(args)
         self.tdm_params = args
         self.shift = shift
         self.loop_vars = self.params(*[f"p{i}" for i in range(len(args))])
         return self
 
+    # pylint: disable=too-many-branches
+    def compile(self, *, device=None, compiler=None):
+        """Compile the time-domain program given a Strawberry Fields photonic hardware device specification.
+        At this stage the compilation is simply a check that the program matches the device.
+
+        Args:
+            device (~strawberryfields.api.DeviceSpec): device specification object to use for
+                program compilation
+            compiler (str, ~strawberryfields.compilers.Compiler): Compiler name or compile strategy
+                to use. If a device is specified, this overrides the compile strategy specified by
+                the hardware :class:`~.DeviceSpec`.
```

Comment:
It might be worth including a list of supported compiler names here (currently, `"gaussian"` and `"TD2"`.